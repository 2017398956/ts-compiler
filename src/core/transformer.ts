import ts from 'typescript/lib/typescript';

/**
 * 提供一个transform helper函数
 */
export function visitNodes(
  node: ts.Node,
  context: ts.TransformationContext,
  callback: (node: ts.Node, context: ts.TransformationContext) => ts.Node
): ts.SourceFile
export function visitNodes(
  node: ts.Node,
  context: ts.TransformationContext,
  callback: (node: ts.Node, context: ts.TransformationContext) => ts.Node
): ts.Node {
  const newNode = callback(node, context);
  if (node !== newNode) {
    return newNode;
  }
  return ts.visitEachChild(node, childNode => visitNodes(childNode, context, callback), context);
}

export const createTransformer = (callback: (node: ts.Node, context: ts.TransformationContext) => ts.Node): ts.TransformerFactory<ts.SourceFile> =>
  context => node => visitNodes(node, context, callback)