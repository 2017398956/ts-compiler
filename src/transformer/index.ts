export * from './noexcept'
export * from './transformer'
