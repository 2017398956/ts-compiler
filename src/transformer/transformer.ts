import ts from 'typescript/lib/typescript'
import { createAstNode } from '../traverser'

/**
 * 提供一个transform helper函数
 */
export function visitNodes(
  node: ts.Node,
  context: ts.TransformationContext,
  callback: (node: ts.Node, context: ts.TransformationContext) => ts.Node
): ts.SourceFile
export function visitNodes(
  node: ts.Node,
  context: ts.TransformationContext,
  callback: (node: ts.Node, context: ts.TransformationContext) => ts.Node
): ts.Node {
  const newNode = callback(node, context)
  if (node !== newNode) {
    return newNode
  }
  return ts.visitEachChild(
    node,
    childNode => visitNodes(childNode, context, callback),
    context
  )
}

/**
 * 创建一个ast转换器
 */
export const createTransformer =
  (
    callback: (node: ts.Node, context: ts.TransformationContext) => ts.Node
  ): ts.TransformerFactory<ts.SourceFile> =>
  context =>
  node =>
    visitNodes(node, context, callback)

export const transform = (
  text: string,
  transformers: ts.TransformerFactory<ts.SourceFile>[],
  scriptKind: ts.ScriptKind = ts.ScriptKind.TSX
) => {
  const ast = createAstNode(text, scriptKind)
  const newAst = ts.transform(ast, transformers)
  const printer = ts.createPrinter()
  return printer.printNode(ts.EmitHint.SourceFile, newAst.transformed[0], ast)
}
