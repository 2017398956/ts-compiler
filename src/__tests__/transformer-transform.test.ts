import { createTransformer } from '../transformer/transformer'
import { transform } from '../transformer'
import ts from 'typescript/lib/typescript'

describe('Transform', () => {
  it('transform', () => {
    const code = `
interface Test {
  name: string;
}
`

    const result = transform(code, [
      createTransformer(node => {
        // 找到属性
        if (ts.isPropertySignature(node)) {
          return ts.addSyntheticLeadingComment(
            node,
            ts.SyntaxKind.MultiLineCommentTrivia,
            `this is a jsdoc comment`,
            true
          )
        }
        return node
      }),
    ])

    expect(result).toMatchSnapshot()
  })
})
