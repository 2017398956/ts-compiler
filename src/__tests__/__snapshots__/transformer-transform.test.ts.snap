// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transform transform 1`] = `
"interface Test {
    /*this is a jsdoc comment*/
    name: string;
}
"
`;
